### ISSUE TITLE
horribly high system load

### ISSUE THREAD

**Issue by nonchip** [ValveSoftware/steam-for-linux#6073]
#### Your system information

* Steam client version (build number or date): current beta
* Distribution (e.g. Ubuntu): void
* Opted into Steam client beta?: Yes
* Have you checked for system updates?: Yes

#### Please describe your issue in as much detail as possible:

whenever steam downloads/updates a game and accesses the disk (as seen both in steam's graph and htop's "DISK R/W" column) it hogs it so bad everything freezes up from iowait. loadavg shoots up to above 100.

#### Steps for reproducing this issue:

1. download something in steam
2. watch the io load or try watching a youtube vid, or even moving the mouse around (anything that needs the slightest amount of I/O time)
3. go crazy because your system is totally unresponsive

in case you need any logs/etc, please do tell

**Comment by aconscious** [comment]
I'm also experiencing this issue. 
**Your system information**

- Steam client version (build number or date): Also current Beta
-     Distribution (e.g. Ubuntu): Arch Linux
-     Opted into Steam client beta?: Yes
-     Have you checked for system updates?: Yes

My setup also runs a LUKS encrypted LVM set which I thought might be related, since steam both has to decrypt and/or decompress the files and then lets the kernel encrypt them when writing.

I have been able to slightly mitigate the issue by throttling the download speed to 3MB/s, but the system is still severly crippled by the downloads.

Other things I have tinkered with:
- Setting different IO-schedulers (none, mq-deadline, Kyber, BFQ)
- Experimenting with allow-discard boot parameter (TRIM)
- Experimenting with sysctl settings (swappiness, vm.dirty(_background)_ratio, vm.vfs_cache_pressure)
- Experimenting with installing to separate LUKS/LVM devices not part of the / (root) or /home device

Noteworthy is this only seems to happen when High IO is created by the steam client when downloading/installing games

**Comment by smirgol** [comment]
I'm recently also having a very similar issue. I don't know exactly when it started, but it cannot be that long ago, a couple of weeks maybe.
What happens is, that sometimes, but not always, when I start a game and it is starting "Processing Vulkan shaders", it gets stuck.
I/O is at 100% and nothing is going to happen for up to 30 minutes or even longer. CPU load is normal, it's just the disk I/O. I'm not having the issue that my system becomes unresponsive though, it's just the game won't start until steam has finished whatever it is doing.

I have absolutely no idea what's going on and why it takes that long. At some point, which can really take a long time, it finishes, compiles the shaders and everything is good - until next time. I think it also doesn't happen all the time when shaders need to be rebuilt and, not sure about that, it also happens without building shaders. I know, very its pretty confusing, so it is to me.

Is there any way I can debug what is going on? I find it very difficult to get a hold on the detailed I/O information.
All that I can see in iotop is:
`steam -nominidumps -nobreakpad [CJobMgr::m_Work]`

- Steam client version (build number or date): Current Beta
- Distribution (e.g. Ubuntu): Ubuntu 20.04
- Opted into Steam client beta?: Yes
- Have you checked for system updates?: Yes

[Steam system information](https://gist.github.com/smirgol/59c882fdb1a1497b1cce5ee7d4158436)

**Comment by pablow1422** [comment]
Same issue here. Although I don't lose responsiveness (since my steam library resides in another HDD), I'm also seeing that high I/O when launching a game.

- Steam client version (build number or date): 163992987
- Distribution (e.g. Ubuntu): openSUSE Tumbleweed 20201203
- Opted into Steam client beta?: No
- Have you checked for system updates?: Yes

**Comment by h1z1** [comment]
That would be a kernel function. Is system tracing enabled ?  

```
grep . /sys/kernel/debug/tracing/{tracing_on,events/ext4/ext4_mark_inode_dirty/enable}

```
Check syslog or /proc/kmsg

One of the things Steam tries to do is enable it.  The client used to complain if it couldn't.    Problem isn't necessarily with tracing itself but the level of tracing Valve enables is .. silly.  It will slam the hell out of syslog and was a basis for them requiring root since you can't enable it as a user (nor should it ever outside a DEVEL environment).

tl;dr - Steam writing it out to a log can get amplified very quickly.   The write itself can cause another event which snowballs into a loop.  It gets worse when you add on anything else logging klog like syslogd or worse journald.

**Comment by DanielGaaA** [comment]
Lately this was getting really bad. Freezes while finalizing download could cause Firefox to crash(before it was just desktop freeze for few seconds). I have slow QLC nvme drive that I though was TLC when buying.... Thanks Crucial for sending reviewers TLC drives while selling it as QLC drive. Don't buy Crucial P2

Anyway all my problems went away when I switched to "Kyber" I/O scheduler.

**Comment by nonchip** [comment]
yup we're there again, especially after boot it catapults my load (on a ryzen5 with ssd!) up to >10 consistently, by deciding to upgrade a bunch of wineprefixes for no reason while apparently unpacking a throttled download with a thread per core.

my problem definitely wasn't kernel tracing or extfs related, given i didn't use extfs since opening this, and steam never runs as root, who gave you that idea @h1z1 :P

**Comment by nonchip** [comment]
ok seriously this is ridiculous, can *literally* anyone at valve care?

![Image](https://github.com/user-attachments/assets/e650cf91-d744-4505-995a-cd08f282e277)

**Comment by n00dl3** [comment]
Tried every solution, responsiveness is way better but I'm stuck at 30MB/s -at best- on an NVMe SSD. A Helldivers 2 upgrade takes litteraly HOURS. This is driving me crazy.

**Comment by nonchip** [comment]
can confirm @n00dl3 some games will still get the load absurdly high and be stuck at low actual throughput. Helldivers 2 is one of the worst offenders there. that said, plenty people on all platforms complain about HD2 updates regularly, it's not uncommon for a 100 byte download to cause the whole game to be rewritten on disk for no apparent reason.

**Comment by nonchip** [comment]
@kisak-valve since you have clearly been aware of this ongoing issue since at least 2021, is there **any** chance that you guys give us a simple input/checkbox to limit the amount of threads? it's getting ridiculous.

**Comment by Askaiant** [comment]
i can confirm Helldivers 2 updates/installation is just the worst. The problem is even when you pause the installation. It will take some time until the last chunk in unpacked, however it seems that it doesn't release the I/O from my SSD. So even shutting down the PC becomes a half hour long process. 

I am happy to provide input. This issue is really killing us.

**Comment by tsokiyZan** [comment]
want to pop in here and say that this is still an issue, on arch and my poor SSD get absolutely HAMMERED every single time I open steam without fail causing my whole system to grind to a halt

**Comment by gornvan** [comment]
Here to confirm the problem still exists. While Steam is doing its downloads and patches, the whole system becomes completely unusable.

UPD:
Some mangling with `nice` and `ionice` allowed to tame this beast by pushing its priority down relatively to the rest of the system. All the GUIs are now responsive as they are normally. Steam now unpacks the updates horribly slowly, but that's a trade I'm agreeing to.
